pkg/operator/controller/dnsnameresolver/resolver.go

```diff
@@ -0,0 +1,340 @@
+	defaultPort      = "53"
+	maxCoreDNSPodIPs = 5
+)
+
+// dnsResolvedName contains details about a DNS resolved name.
+type dnsResolvedName struct {
+	minNextLookupTime time.Time
+	regularObjExists  bool
+	wildcardObjExists bool
+	numIPs            int
+}
+
+// Resolver keeps track of the DNS names, the corresponding IP addresses and TTLs. It sends DNS resolution
+// requests to CoreDNS pods whenever TTL of any IP address associated to a DNS name expires.
+type Resolver struct {
+	corednsEndpointSliceCache cache.Cache
+	serviceName               string
+	dnsNames                  map[string]dnsResolvedName
+	regularObjInfo            map[string]string
+	wildcardObjInfo           map[string]sets.Set[string]
+	dnsLock                   sync.Mutex
+
+	added   chan struct{}
+	deleted chan string
+}
+
+// NewResolver initializes and returns a new resolver instance.
+func NewResolver(cache cache.Cache, serviceName string) *Resolver {
+	return &Resolver{
+		corednsEndpointSliceCache: cache,
+		serviceName:               serviceName,
+		dnsNames:                  make(map[string]dnsResolvedName),
+		regularObjInfo:            make(map[string]string),
+		wildcardObjInfo:           make(map[string]sets.Set[string]),
+		added:                     make(chan struct{}, 1),
+		deleted:                   make(chan string, 1),
+	}
+}
+
+// Start starts a goroutine for sending DNS lookup requests of the DNS names on expiration of the corresponding TTLs.
+func (res *Resolver) Start() {
+	var nextDNSName, deletedDNSName string
+	var nextLookupTime time.Time
+	var exists bool
+	var numIPs int
+	// Initially the check for nextDNSName to be looked up will happen after default maximum TTL. If a DNS name
+	// gets added before that then the nextDNSName and timeTillNextLookup will be updated.
+	timeTillNextLookup := defaultMaxTTL
+	go func() {
+		timer := time.NewTicker(timeTillNextLookup)
+		defer timer.Stop()
+		for {
+			select {
+			case <-res.added:
+				// Whenever DNS name(s) are added or updated, the nextDNSName and timeTillNextLookup
+				// needs to be updated.
+			case <-timer.C:
+				// After every tick, DNS lookup needs to be performed for nextDNSName, if it is not empty.
+				// The nextDNSName and timeTillNextLookup also needs to be updated.
+				if len(nextDNSName) > 0 {
+					if err := res.lookupDNSName(nextDNSName, numIPs); err != nil {
+						logrus.Warning(err)
+					}
+				}
+			case deletedDNSName = <-res.deleted:
+				// If the deleted DNS name is the DNS name that will be looked up next, then
+				// the nextDNSName and timeTillNextLookup needs to be updated. If another
+				// DNS name is deleted then nothing is needed to be done.
+				if nextDNSName != deletedDNSName {
+					continue
+				}
+			}
+			// Get the DNS name whose TTL will expire first. Based on the nextLookupTime for the nextDNSName, the timeTillNextLookup
+			// will be updated appropriately.
+			nextDNSName, nextLookupTime, numIPs, exists = res.getNextDNSNameDetails()
```

Comment:
This may deadlock, because we're in a loop and the function needs to wait for access to `res`.  Have you done any kind of testing to make sure there are no deadlocks?

Reply:
The `dnsLock` mutex of `res` is used in Add, Deleted and getNextDNSNameDetails functions to serialize the access to the shared maps. As a single mutex variable is used no deadlock will arise.

Deadlock may happen if there were two or more mutex variables and one of the functions have taken a lock on one mutex and waiting on another mutex which have been locked by another function, and this function is waiting to take a lock on the first mutex.

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2